Add IssuedUtc and ExpiresUtc fields to OpenIddictToken

vyfster opened the pull request:
Fields (`IssuedUtc` and `ExpiresUtc`) added to help keep track of tokens added to the `OpenIddictTokens` table so that expired tokens can be easily identified.

Unit tests have been updated with the change, however the tests `SerializeAuthorizationCode_AuthorizationCodeIsCorrectlyPersisted` and `SerializeRefreshToken_RefreshTokenIsCorrectlyPersisted` should probably include a check the verifies that the values that have been persisted for these two fields is correct.  For this to happen, the `AuthenticationTicket` that is created in `OpenIddictProviderTests` line 142 needs to be available to the unit tests.  I'm not sure on the best approach to achieve this - any hints / guidelines on doing this?

*** NOTE - this breaks if used with an existing database that has not been updated.  Two new columns must be added to the table `OpenIddictTokens`.  Alternatively, if the data in the database is not needed, as it's in dev, then simply drop the database.

**MSSQL**
IssuedUtc DateTimeOffset (NULL)
ExpiresUtc DateTimeOffset (NULL)

**MySQL || MariaDB**
IssuedUtc DATETIME (NULL)
ExpiresUtc DATETIME (NULL)

vyfster committed:
Add IssuedUtc and ExpiresUtc fields to OpenIddictTokens

 - fields to help keep track of tokens added to the OpenIddictTokens
 table so that expired tokens can be deleted from the table.

kevinchalet commented:
> For this to happen, the AuthenticationTicket that is created in OpenIddictProviderTests line 142 needs to be available to the unit tests. I'm not sure on the best approach to achieve this - any hints / guidelines on doing this?

There's a simple way to do that: after line 142, add `ticket.Properties.IssuedUtc = new DateTimeOffset(... fixed date...)`, which will force OpenIddict to use a fixed date instead of the current date time. Then, call `ticket.SetAuthorizationCodeLifetime(TimeSpan.FromDays(1))` (or `ticket.SetRefreshTokenLifetime(TimeSpan.FromDays(1))`) to set the expiration date to `ticket.Properties.IssuedUtc` + 1 day, and update your tests to use your hardcoded dates instead of `It.IsAny<DateTimeOffset?>()`.

kevinchalet commented:
Note: I'll probably wait a bit before merging your PR. I'd like to combine it with https://github.com/openiddict/openiddict-core/issues/433 - which will require massive changes in the `OpenIddictToken` entity - so we avoid breaking people twice :sweat_smile:

vyfster commented:
> There's a simple way to do that: after line 142, add ticket.Properties.IssuedUtc = new DateTimeOffset(... fixed date...), which will force OpenIddict to use a fixed date instead of the current date time.

Unfortunately this doesn't seem to be the case.  In the file `OpenIddictProvider.Serialization.cs`, method `SerializeAuthorizationCode`, line 37, the value for `context.Ticket.Properties.IssuedUtc` seems to be set to DateTime.UtcNow (or similar value), regardless of whether the value for this property has been set.

I've had a dig through the AspNet.Security.OpenidConnect.Server code and it looks like the `IssuedUtc` property is set there (code below is present in multiple places):
```c#
            // Create a new ticket containing the updated properties.
            var ticket = new AuthenticationTicket(identity, properties);
            ticket.Properties.IssuedUtc = Options.SystemClock.UtcNow;
```

I'm not familiar with the code so I might be wrong wrt ASOS, but IssuedUtc is always set to a utc date regardless of whether or not it has a value.

Simple way to test - set to a hardcoded value as described in the comments above and then set a breakpoint at line 37 in `OpenIddictProvider.Serialization.cs`.

kevinchalet commented:
Arg yeah, that's true (I now remember I changed that at some point, when I decided to ignore `ExpiresUtc` and replace it by the `ticket.SetXYZLifetime()` extensions).

Another approach is to use a mocked clock to ensure a fixed date is used for `IssuedUtc`:

E.g:

```csharp
var server = CreateAuthorizationServer(builder =>
{
    builder.Configure(options =>
    {
        options.SystemClock = Mock.Of<ISystemClock>(mock => mock.UtcNow ==
            new DateTimeOffset(2017, 01, 01, 00, 00, 00, TimeSpan.FromHours(0)));
    });

    builder.Services.AddSingleton(manager);
});
```

vyfster committed:
Fix unit tests to verify correct values for ticket dates.

 - Use a mocked system clock to ensure a fixed date is used for
 IssuedUtc instead of DateTime.UtcNow, for unit tests.

 - Replaced It.IsAny<DateTimeOffset> in mock setup with expected values to
 verify that expected date values are passed in as arguments to the CreateAsync method.

kevinchalet commented:
Your commit has been integrated into https://github.com/openiddict/openiddict-core/pull/443 and backported to OpenIddict 1.x.

Thanks for your contribution! :clap:
